File: C++/207. Course Schedule.cpp
#include <bits/stdc++.h>

using namespace std;

bool canFinish(int numCourses, vector<vector<int>> &prerequisites)
{
    vector<vector<int>> graph(numCourses, vector<int>()); // Adjacency List
    vector<int> remaining(numCourses, 0);

    for (int i = 0; i < prerequisites.size(); i++)
    {
        graph[prerequisites[i][1]].push_back(prerequisites[i][0]);
        remaining[prerequisites[i][0]]++;
    }

    queue<int> que;

    for (int i = 0; i < numCourses; i++)
    {
        if (remaining[i] == 0)
            que.push(i);
    }

    int course;

    while (!que.empty())
    {
        course = que.front();
        que.pop();

        for (int higher_course : graph[course])
        {
            remaining[higher_course]--;
            if (remaining[higher_course] == 0)
                que.push(higher_course);
        }
    }

    for (int i = 0; i < numCourses; i++)
    {
        if (remaining[i])
            return false;
    }

    return true;
}

File: Java/207. Course Schedule.java

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = new ArrayList<>(numCourses);
        int[] incomming = new int[numCourses];

        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] req : prerequisites) {
            graph.get(req[1]).add(req[0]);
            incomming[req[0]]++;
        }

        Queue<Integer> queue = new LinkedList<>();

        for (int i = 0; i < numCourses; i++) {
            if (incomming[i] == 0)
                queue.add(i);
        }

        while (!queue.isEmpty()) {
            int preCourse = queue.poll();

            for (int course : graph.get(preCourse)) {
                if (--incomming[course] == 0)
                    queue.add(course);
            }
        }

        for (int course : incomming)
            if (course > 0)
                return false;

        return true;
    }
}